In [4]:
import pandas as pd
import numpy as np

In [5]:
pd.set_option('display.max_columns', None)

# inlezen van data

in de data zijn values in een column met verschillende Type, met low_memory=False kunt pandas bij inlezen een faste Type voor een column vast leggen.

In [6]:
df = pd.read_csv('sap_storing_data_hu_project.csv', low_memory=False)

haal alle collumn dat ? of ntv zijn

In [7]:
df = df[["stm_mon_nr","stm_vl_post","stm_sap_meld_ddt","stm_sap_meldtekst_lang","stm_sap_meldtekst","stm_geo_mld","stm_equipm_nr_mld",
                    "stm_equipm_soort_mld","stm_equipm_omschr_mld","stm_km_van_mld","stm_km_tot_mld","stm_prioriteit","stm_aanngeb_ddt","stm_oh_pg_gst",
                    "stm_geo_gst","stm_equipm_nr_gst","stm_equipm_soort_gst","stm_equipm_omschr_gst","stm_km_van_gst","stm_km_tot_gst","stm_oorz_groep",
                    "stm_oorz_code","stm_oorz_tkst","stm_fh_ddt","stm_fh_status","stm_sap_storeind_ddt","stm_tao_indicator","stm_tao_indicator_vorige",
                    "stm_tao_soort_mutatie","stm_tao_telling_mutatie","stm_tao_beinvloedbaar_indicator","stm_sap_melddatum","stm_sap_meldtijd","stm_contractgeb_mld",
                    "stm_techn_mld","stm_contractgeb_gst","stm_techn_gst","stm_aanngeb_dd","stm_aanngeb_tijd","stm_aanntpl_dd","stm_aanntpl_tijd","stm_progfh_in_datum",
                    "stm_progfh_in_tijd","stm_progfh_in_invoer_dat","stm_progfh_in_invoer_tijd","stm_progfh_in_duur","stm_progfh_gw_tijd","stm_progfh_gw_teller",
                    "stm_fh_dd","stm_fh_tijd","stm_fh_duur","stm_sap_storeinddatum","stm_sap_storeindtijd","stm_oorz_tekst_kort","stm_pplg_van","stm_pplg_naar",
                    "stm_dstrglp_van","stm_dstrglp_naar"]]

het data bastaan voraal uit drie sorten type data:<br>
tekstuele inhoud<br>
numerike classen<br>
datum en tijd<br>

In [8]:
df 

stm_mon_nr stm_vl_post     stm_sap_meld_ddt  \
0              NaN         NaN                  NaN   
1              0.0         NaN  02/01/2006 09:00:00   
2             48.0          GN  02/01/2006 12:35:00   
3             72.0          ZL  02/01/2006 16:40:00   
4             96.0          ZL  02/01/2006 22:30:00   
...            ...         ...                  ...   
898521         0.0         NaN  11/05/2013 08:06:00   
898522         0.0         NaN  11/05/2013 09:21:00   
898523         0.0         NaN  11/05/2013 09:21:00   
898524         0.0         NaN  20/08/2016 14:15:17   
898525         0.0         NaN  20/08/2016 14:15:17   

                                   stm_sap_meldtekst_lang  \
0                                                     NaN   
1        Logboeknr  Geeltje  :   49 Tijd: 0900   VL-Po...   
2        Logboeknr  RBV  :   48 Tijd: 1235   VL-Post: ...   
3        Logboeknr  RBV  :   72 Tijd: 1640   VL-Post: ...   
4        Logboeknr  RBV  :   96 Tijd: 2230   VL-Post: ...   
...                                                   ...   
898521   GVC : sp-2 en 3  stop lampen op de stootjukke...   
898522   Whz : Ovw 100.1 albert Plesmanweg blijft  dic...   
898523   Whz : Ovw 100.1 albert Plesmanweg blijft  dic...   
898524  Bdg-Apn : Km 12.7 twee gedode zwanen langs het...   
898525  Bdg-Apn : Km 12.7 twee gedode zwanen langs het...   

                               stm_sap_meldtekst stm_geo_mld  \
0                                            NaN         NaN   
1          Logboeknr  Geeltje  :   49 Tijd: 0900       624.0   
2       Logboeknr  RBV  :   48 Tijd: 1235   VL-P       201.0   
3       Logboeknr  RBV  :   72 Tijd: 1640   VL-P        25.0   
4       Logboeknr  RBV  :   96 Tijd: 2230   VL-P        12.0   
...                                          ...         ...   
898521  GVC : sp-2 en 3  stop lampen op de stoot       560.0   
898522  Whz : Ovw 100.1 albert Plesmanweg blijft       468.0   
898523  Whz : Ovw 100.1 albert Plesmanweg blijft       468.0   
898524  Bdg-Apn : Km 12.7 twee gedode zwanen lan       102.0   
898525  Bdg-Apn : Km 12.7 twee gedode zwanen lan       102.0   

        stm_equipm_nr_mld stm_equipm_soort_mld  \
0                     NaN                  NaN   
1                     NaN                  NaN   
2                     NaN                  NaN   
3                     NaN                  NaN   
4                     NaN                  NaN   
...                   ...                  ...   
898521                NaN                  NaN   
898522         10021085.0              OVERWEG   
898523         10021085.0              OVERWEG   
898524                NaN                  NaN   
898525                NaN                  NaN   

                           stm_equipm_omschr_mld  stm_km_van_mld  \
0                                            NaN             NaN   
1                                            NaN             0.0   
2                                            NaN             0.0   
3                                            NaN             0.0   
4                                            NaN             0.0   
...                                          ...             ...   
898521                                       NaN         24500.0   
898522  Havio 100.1 A Rtd Eemh. Albert Plesmanwe        100051.0   
898523  Havio 100.1 A Rtd Eemh. Albert Plesmanwe        100051.0   
898524                                       NaN             0.0   
898525                                       NaN             0.0   

        stm_km_tot_mld  stm_prioriteit      stm_aanngeb_ddt stm_oh_pg_gst  \
0                  NaN             NaN                  NaN           NaN   
1                  0.0             9.0  02/01/2006 09:00:00           81S   
2                  0.0             9.0  02/01/2006 12:35:00           37B   
3                  0.0             9.0  02/01/2006 16:40:00           32B   
4                  0.0             9.0  02

Tijdens de proces wanneer het model gebruik wordt zijn niet alle data zijn beschikbaar. Data dat eindig met _gst zijn niet geschikte data.

# filter doubble data

In [49]:
df = df.drop_duplicates()

# feature

In [148]:
data=df[['stm_oorz_code','stm_sap_melddatum','stm_sap_meldtijd','stm_geo_mld','stm_aanntpl_tijd','stm_fh_tijd','stm_techn_mld','stm_prioriteit','stm_contractgeb_mld','stm_fh_duur','stm_progfh_in_duur','stm_progfh_in_tijd']]

hier zijn wat feature tijdens proces beschiktbaar is en wat we intersant vinden:<br>
stm_oorz_code : code van wat de probleem is (weet pas was de aannemer aanwezig is) <br>
stm_sap_melddatum: de datum data melding melding is binnen gekommen <br>
stm_sap_meldtijd : de tijdstip van de melding <br>
stm_geo_mld: geocode waar iets gebuurt is. <br>
stm_aanntpl_tijd : de tijdmoment dat de aanmener anweizig is <br>
stm_fh_dd : de datum dat de probleem opgelost is <br>
stm_aanntpl_tijd : de tijdmoment dat de probleem opgelost is （het is binnen 8uur het kan zijn dat dit niet van belang is） <br>
stm_techn_mld: Bij welke techniekveld de melding hoort<br>
stm_prioriteit: De prioriteit zeg over hoebelangrijk en hoe snel de probleem zijn <br>
stm_fh_duur: hoelang het duurt van melding tot dat de probleem is opgelost
stm_progfh_in_tijd: tijdstip waarneer het klaar zijn. 

# FHT prognose 

eerts kijk we wat stm_fh_tijd presis betekend<br>
Om beter mee te rekenen zet we de tijd om in hoeveelste minuut van de dag.

In [149]:
# Kolomnamen met tijd in HH:MM:SS formaat
tijd_kolommen = ['stm_sap_meldtijd', 'stm_aanntpl_tijd', 'stm_fh_tijd','stm_progfh_in_tijd']

# Functie om tijdstring om te zetten naar totaal aantal minuten
def convert_time_to_minutes(df, columns):
    for col in columns:
        df.loc[:, col] = pd.to_datetime(df[col], format='%H:%M:%S', errors='coerce')
        df.loc[:, col] = df[col].apply(lambda x: x.hour * 60 + x.minute if pd.notnull(x) else None)
    return df


In [150]:
data = convert_time_to_minutes(data, tijd_kolommen)

In [151]:
data[['stm_sap_meldtijd', 'stm_aanntpl_tijd', 'stm_fh_tijd','stm_fh_duur']]

stm_sap_meldtijd stm_aanntpl_tijd stm_fh_tijd  stm_fh_duur
0                   NaN              NaN         NaN          NaN
1                 540.0              0.0       540.0          0.0
2                 755.0              0.0       806.0         51.0
3                1000.0              0.0      1040.0         40.0
4                1350.0              0.0      1356.0          6.0
...                 ...              ...         ...          ...
898521            486.0            545.0       569.0         83.0
898522            561.0            608.0       644.0         83.0
898523            561.0            608.0       644.0         83.0
898524            855.0            885.0       914.0         19.0
898525            855.0            885.0       914.0         19.0

[831507 rows x 4 columns]

Hieruit blijkt dat functie hersteel duur gelijk aan hersteel tijdstip - meldings tijdstip<br>
ook zien je bij somig storing dat geen infomatie is wanneer aannemer aanweizig was.

In [152]:
len(data[data['stm_fh_duur'] == 0])

160611

In [153]:
len(data.loc[(data["stm_progfh_in_tijd"] == data["stm_fh_tijd"]) & (df["stm_fh_tijd"] != 0)])

291345

In [154]:
len(data[data['stm_fh_duur'] == (data['stm_progfh_in_tijd']-data['stm_aanntpl_tijd'])])

156596

Hieruit kunnen we zien dat 160.000 data geen stroning bij melding.<br>
140.000 data prognose functie herstel tijdstip gelijk zijn aan werkelijk hersteltijd<br>
150.00 data zijn werkelijk FHT gelijk aan tijdstip aannemer te plaats tot prognose functie herstel tijdstip 

in stm_progfh_in_duur zijn str values, we bewaar eerste alle numeric waardes en zet we de type als int

In [155]:
data.loc[:, 'stm_progfh_in_duur'] = pd.to_numeric(data['stm_progfh_in_duur'], errors='coerce')
data.loc[:, 'stm_progfh_in_duur'] = data['stm_progfh_in_duur'].fillna(0).astype(int)

In [156]:
len(data[data['stm_progfh_in_duur'] == (data['stm_progfh_in_tijd']-data['stm_aanntpl_tijd'])])

486618

In [157]:
len(data[data['stm_progfh_in_duur'] == data['stm_fh_duur']])

95215

verder is te zien dat, 120.000 data dat prognose functie herstel tijd gelijk tijdstip aannemer te plaats + prognose functie herstel tijdstip <br>
ook zijn 9.000 data dat prognose duur gelijk aan werkelijk FHT duur

# Target

voor de traget neem we de tijd neem de functie hersteltijd van waarneer de aanmener aanweizig is tot dat weer gereiden mogen worden.<br>
we bewaar eerst alle data dat binnen dag is

In [158]:
data['targetherstel'] = np.where(
    data['stm_fh_tijd'] - data['stm_aanntpl_tijd']>= 0, # als de stm_fh_tijd-stm_aanntpl_tijd positief is
    data['stm_fh_tijd'] - data['stm_aanntpl_tijd'], # dan betekende dat nog op de zefde day is, kun je gelijk van elkaar aftrekken
    data['stm_fh_tijd'] - data['stm_aanntpl_tijd']+1440 # als niet op zelfdde dag is hersteld, dus een neagetief uitkomst moet er +1440 om de juiste tijd te krijgen.
)

C:\Users\maste\AppData\Local\Temp\ipykernel_8992\3034723497.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['targetherstel'] = np.where(


In [159]:
ranges = {
    "null": (0, 0),
    "tot 15m": (0, 15),
    "15m tot 30m": (15, 30),
    "30m tot 1h": (30, 60),
    "1h tot 2h": (60, 120),
    "2h tot 3h": (120, 180),
    "3h tot 6h": (180, 360),
    "6h tot 8h": (360, 540),
    "8h+": (540, float('inf'))
}

total = len(data)

for label, (low, high) in ranges.items():
    if label == "null":
        count = len(data[data['targetherstel'] == 0])
    else:
        count = len(data[(data['targetherstel'] > low) & (data['targetherstel'] <= high)])
    print(f'score {label}: {count/total:.2%}')


score null: 20.53%
score tot 15m: 21.31%
score 15m tot 30m: 12.51%
score 30m tot 1h: 15.28%
score 1h tot 2h: 12.22%
score 2h tot 3h: 4.81%
score 3h tot 6h: 5.22%
score 6h tot 8h: 1.74%
score 8h+: 6.38%


we zien dat dat 20% van de meldingen geen probleem zijn dat de spoor op stil staat
21% dat onder de 15 miuten is en 13% tussen 15-30 min
voor meldingen dat de spoor boven de acht uurr stil staan zij er maar 6%

In [160]:
len(data[data['stm_progfh_in_duur'] == data['targetherstel']])

196142

# Filter

filter data's dat stm_fh_duur onder de 15 min en data's boven de 8 uur,het is niet belangrjik in deze project.<br>
we nemen data's van 15min tot 8uur, zo data de model niet telaag gaat onderschaten.

In [175]:
data = data.loc[ (data['targetherstel']>15) & (data['targetherstel']<=540)]

Filter van grognose duur, er mag geen rare waardes als 0 of 99999999 in staan

In [176]:
data = data.loc[(data['stm_progfh_in_duur']>0) & (data['stm_progfh_in_duur']<=1440)]

Filter voor menselijk fouten waar de waarders achteraf is door de aannemer ingevoerd zijn.

In [177]:
data = data.loc[(data["stm_progfh_in_tijd"] != data["stm_fh_tijd"])]

In [178]:
data = data[data['stm_progfh_in_duur'] != data['stm_fh_duur']]

In [179]:
data = data[data['stm_progfh_in_duur'] != data['targetherstel']]

# datum reparatie

In [166]:
data = data.copy()
#Zet de kolom om naar datetime,vervang de datum door de dag van het jaar
data.loc[:, 'stm_sap_melddatum'] = pd.to_datetime(data['stm_sap_melddatum'], format='%d/%m/%Y').dt.dayofyear

# prioriteit

In [167]:
data=data.loc[(data['stm_prioriteit']!=8)&(data['stm_prioriteit']!=9)]

In [168]:
data

stm_oorz_code stm_sap_melddatum stm_sap_meldtijd stm_geo_mld  \
95714           215.0               201            545.0       119.0   
95715           215.0               201            545.0       119.0   
139811          215.0               279            749.0        63.0   
139812          215.0               279            749.0        63.0   
139816          215.0               279           1012.0        62.0   
...               ...               ...              ...         ...   
898521          215.0               131            486.0       560.0   
898522          218.0               131            561.0       468.0   
898523          218.0               131            561.0       468.0   
898524          135.0               233            855.0       102.0   
898525          143.0               233            855.0       102.0   

       stm_aanntpl_tijd stm_fh_tijd stm_techn_mld  stm_prioriteit  \
95714              30.0       240.0           NaN             2.0   
95715              30.0       240.0           NaN             2.0   
139811            788.0       810.0             S             4.0   
139812            788.0       810.0             S             4.0   
139816            583.0       632.0             S             4.0   
...                 ...         ...           ...             ...   
898521            545.0       569.0           NaN             5.0   
898522            608.0       644.0             S             2.0   
898523            608.0       644.0             S             2.0   
898524            885.0       914.0           NaN             2.0   
898525            885.0       914.0           NaN             2.0   

        stm_contractgeb_mld  stm_fh_duur stm_progfh_in_duur  \
95714                   NaN       1012.0                270   
95715                   NaN       1012.0                270   
139811                 26.0         53.0                 31   
139812                 26.0         53.0                 31   
139816                 25.0       1051.0                 41   
...                     ...          ...                ...   
898521                  NaN         83.0                 30   
898522                  4.0         83.0                 52   
898523                  4.0         83.0                 52   
898524                  NaN         19.0                 75   
898525                  NaN         19.0                 75   

       stm_progfh_in_tijd targetherstel  
95714               300.0         210.0  
95715               300.0         210.0  
139811              819.0          22.0  
139812              819.0          22.0  
139816              624.0          49.0  
...                   ...           ...  
898521              575.0          24.0  
898522              660.0          36.0  
898523              660.0          36.0  
898524              960.0          29.0  
898525              960.0          29.0  

[242010 rows x 13 columns]

# LinearRegression

In [169]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_squared_error
import joblib

## baseline

We neem RMSE van prognose tijd duur met de traget als de baseline voor de Linearrregressie model.

In [170]:
np.sqrt(mean_squared_error(data['stm_progfh_in_duur'], data['targetherstel']))

198.37210509252483

Naast de RMSE waarde, wanneer kan we zeggen of een verspelling ook een goede verspelling is?<br> Als de verspelling target herstel tijd kan bedekken met een maximale verschil van 15 minuten.

In [233]:
tolerance = 15

correct = (data['stm_progfh_in_duur'] > data['targetherstel']) & \
          ((data['stm_progfh_in_duur'] - data['targetherstel']) <= tolerance)

accuracy = np.mean(correct)

print(f"Accuracy (overschatting binnen {tolerance} minuten): {accuracy:.2%}")

✅ Accuracy (overschatting binnen 15 minuten): 35.24%


In [230]:
np.abs(data['stm_progfh_in_duur'] - data['targetherstel']) 

95714     60.0
95715     60.0
139811     9.0
139812     9.0
139816     8.0
          ... 
898521     6.0
898522    16.0
898523    16.0
898524    46.0
898525    46.0
Length: 242010, dtype: float64

In [181]:
len(data[data['stm_progfh_in_duur'] > data['targetherstel']])/len(data)

0.8232882938721541

Uit data blijk 35.24% van data van de prognose zal goede verspelling zijn, waarbij 82% van alle data altijd de targetherstel tijd kunt bedekken is.<br>
Het is belangrijk dat de voorspelling beter overschat dan onderschat in de project.

Dus wanner is het model beter dan nu?<br>
1.Als RMSE kleiner wordt<br>
2.Als verspelling vanm  met marge van 15 minuten het Target kunt bedekken<br>
3.Als de totale verspelling een betere dekkingsgraad heeft.

## model

In [182]:
# zet alle Nan naar 0

In [183]:
data = data.fillna(0)

In [242]:
def Train_per_categorical(data):
    data = data.copy()  # Voorkom SettingWithCopyWarning

    # Categorische kolommen die één voor één toegevoegd worden
    categorical_cols = ['stm_oorz_code', 'stm_geo_mld', 'stm_contractgeb_mld', 'stm_techn_mld', 'stm_prioriteit']

    # Numerieke kolommen die altijd in het model zitten
    fixed_numeric_cols = ['stm_sap_meldtijd','stm_sap_melddatum','stm_aanntpl_tijd', 'stm_progfh_in_duur']

    # Eerst de baseline (alleen vaste numerieke features)
    print("Baseline model: alleen numerieke features")
    X_base = data[fixed_numeric_cols]
    y = data['targetherstel']

    X_train, X_test, y_train, y_test = train_test_split(X_base, y, test_size=0.2, random_state=42)
    model = LinearRegression()
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    r2 = r2_score(y_test, y_pred)
    rmse = np.sqrt(mean_squared_error(y_test, y_pred))
    print(f"[Baseline] R² = {r2:.3f}, RMSE = {rmse:.2f}")

    # Accuracy en coverage
    tolerance = 15
    correct = (data['stm_progfh_in_duur'] > data['targetherstel']) & \
              ((data['stm_progfh_in_duur'] - data['targetherstel']) <= tolerance)
    accuracy = np.mean(correct)
    coverage = np.mean(data['stm_progfh_in_duur'] > data['targetherstel'])
    print(f"[Baseline] Accuracy (±{tolerance} min overschatting): {accuracy:.2%}")
    print(f"[Baseline] Dekkingsgraad: {coverage:.2%}")

    # Vervolgens per categorische kolom apart toevoegen
    for col in categorical_cols:
        print(f"\nModel trainen met categorische kolom: '{col}' + vaste numerieke features")

        data[col] = data[col].astype(str)
        X_cat = pd.get_dummies(data[[col]], drop_first=True)

        X = pd.concat([data[fixed_numeric_cols], X_cat], axis=1)

        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

        model = LinearRegression()
        model.fit(X_train, y_train)
        y_pred = model.predict(X_test)

        r2 = r2_score(y_test, y_pred)
        rmse = np.sqrt(mean_squared_error(y_test, y_pred))

        print(f"R² = {r2:.3f}, RMSE = {rmse:.2f}")

        # Accuracy & coverage opnieuw berekenen (zelfde formule)
        correct = (data['stm_progfh_in_duur'] > data['targetherstel']) & \
                  ((data['stm_progfh_in_duur'] - data['targetherstel']) <= tolerance)
        accuracy = np.mean(correct)
        coverage = np.mean(data['stm_progfh_in_duur'] > data['targetherstel'])

        print(f"Accuracy (±{tolerance} min overschatting): {accuracy:.2%}")
        print(f"Dekkingsgraad: {coverage:.2%}")



In [243]:
Train_per_categorical(data)

Baseline model: alleen numerieke features
[Baseline] R² = 0.127, RMSE = 76.48
[Baseline] Accuracy (±15 min overschatting): 35.24%
[Baseline] Dekkingsgraad: 82.33%

Model trainen met categorische kolom: 'stm_oorz_code' + vaste numerieke features
R² = 0.169, RMSE = 74.60
Accuracy (±15 min overschatting): 35.24%
Dekkingsgraad: 82.33%

Model trainen met categorische kolom: 'stm_geo_mld' + vaste numerieke features
R² = -146860698820520.438, RMSE = 991707073.43
Accuracy (±15 min overschatting): 35.24%
Dekkingsgraad: 82.33%

Model trainen met categorische kolom: 'stm_contractgeb_mld' + vaste numerieke features
R² = 0.133, RMSE = 76.19
Accuracy (±15 min overschatting): 35.24%
Dekkingsgraad: 82.33%

Model trainen met categorische kolom: 'stm_techn_mld' + vaste numerieke features
R² = 0.135, RMSE = 76.12
Accuracy (±15 min overschatting): 35.24%
Dekkingsgraad: 82.33%

Model trainen met categorische kolom: 'stm_prioriteit' + vaste numerieke features
R² = 0.130, RMSE = 76.34
Accuracy (±15 min overs